src/hotspot/share/code/codeCache.cpp

```diff
@@ -178,143 +178,144 @@ GrowableArray<CodeHeap*>* CodeCache::_compiled_heaps = new(mtCode) GrowableArray
-      non_nmethod_size += diff_size;
-    }
-  }
+  CodeHeapInfo non_nmethod = {NonNMethodCodeHeapSize, FLAG_IS_CMDLINE(NonNMethodCodeHeapSize), true};
+  CodeHeapInfo profiled = {ProfiledCodeHeapSize, FLAG_IS_CMDLINE(ProfiledCodeHeapSize), true};
+  CodeHeapInfo non_profiled = {NonProfiledCodeHeapSize, FLAG_IS_CMDLINE(NonProfiledCodeHeapSize), true};
 
-  // We do not need the profiled CodeHeap, use all space for the non-profiled CodeHeap
+  const bool cache_size_set   = FLAG_IS_CMDLINE(ReservedCodeCacheSize);
+  const size_t ps             = page_size(false, 8);
+  const size_t min_size       = MAX2(os::vm_allocation_granularity(), ps);
+  const size_t min_cache_size = CodeCacheMinimumUseSpace DEBUG_ONLY(* 3); // Make sure we have enough space for VM internal code
+  size_t cache_size           = ReservedCodeCacheSize;
+
+  // Prerequisites
   if (!heap_available(CodeBlobType::MethodProfiled)) {
-    non_profiled_size += profiled_size;
-    profiled_size = 0;
+    // For compatibility reason, disabled tiered compilation overrides
+    // segment size ever if it was set explicitly.
+    non_profiled.size += profiled.size;
+    // Profiled code heap is not available, forcibly set size to 0
+    profiled.size = 0;
+    profiled.set = true;
+    profiled.enabled = false;
   }
-  // We do not need the non-profiled CodeHeap, use all space for the non-nmethod CodeHeap
+
   if (!heap_available(CodeBlobType::MethodNonProfiled)) {
-    non_nmethod_size += non_profiled_size;
-    non_profiled_size = 0;
+    // Compatibility
+    non_nmethod.size += profiled.size;
+    // Non-Profiled code heap is not available, forcibly set size to 0
+    non_profiled.size = 0;
+    non_profiled.set = true;
+    non_profiled.enabled = false;
+  }
+
+  size_t compiler_buffer_size = 0;
+  COMPILER1_PRESENT(compiler_buffer_size += CompilationPolicy::c1_count() * Compiler::code_buffer_size());
+  COMPILER2_PRESENT(compiler_buffer_size += CompilationPolicy::c2_count() * C2Compiler::initial_code_buffer_size());
+
+  if (!non_nmethod.set) {
+    non_nmethod.size += compiler_buffer_size;
+  }
+
+  if (!profiled.set && !non_profiled.set) {
+    non_profiled.size = profiled.size = (cache_size > non_nmethod.size + 2 * min_size) ?
+                                        (cache_size - non_nmethod.size) / 2 : min_size;
   }
-  // Make sure we have enough space for VM internal code
-  uint min_code_cache_size = CodeCacheMinimumUseSpace DEBUG_ONLY(* 3);
-  if (non_nmethod_size < min_code_cache_size) {
-    vm_exit_during_initialization(err_msg(
-        "Not enough space in non-nmethod code heap to run VM: " SIZE_FORMAT "K < " SIZE_FORMAT "K",
-        non_nmethod_size/K, min_code_cache_size/K));
+
+  if (profiled.set && !non_profiled.set) {
+    non_profiled.size = subtract_size(cache_size, non_nmethod.size + profiled.size, min_size);
+  }
+
+  if (!profiled.set && non_profiled.set) {
+    profiled.size = subtract_size(cache_size, non_nmethod.size + non_profiled.size, min_size);
+  }
```

Comment:
`subtract_size` can do subtracting or nothing by return `min_size`. This is not obvious from its name.
What about a function:
```
// Precondition: either or both of heaps must be set.
//
// If either of heaps size is not set, its size is set to max(available_size - set_heap.size, min_size).
static void set_size_of_unset_code_heap(CodeHeapInfo *heap1, CodeHeapInfo *heap2, size_t available_size, size_t min_size) {
  assert(...); //check precondition
  if (heap1->set && heap2->set)
    return;

  if (!heap2->set)
    swap(heap1, heap2);

  heap1->size = (available_size > heap2->size + min_size) ? (available_size - heap2->size) : min_size;
}
```

Now we can unite two IFs into `else` case of `if (!profiled.set && !non_profiled.set)`:
```
if (!profiled.set && !non_profiled.set) {
 ...
} else {
  set_size_of_unset_code_heap(&profiled, &non_profiled, cache_size -  non_profiled.size, min_size)
}
```

Reply:
Again. The idea is to move to a plain, simple, readable code, if you don't mind

Reply:
> Again. The idea is to move to a plain, simple, readable code, if you don't mind

I don't think `subtract_size` is a good name for the function.
Why does it have three parameters? Logically a subtract operation is `a - b`. A reader of the code will have to go to the definition of the function to understand its logic. This means you can save reader's time just putting the code of `subtract_size` in place or giving more meaningful name.

Reply:
Right. set_size_of_unset_code_heap name is better. Thanks!